### ISSUE TITLE
Documentation rework

### ISSUE THREAD

**Issue by sciurius** [ChordPro/chordpro#390]
Alyn Gwyndaf [wrote](https://groups.io/g/ChordPro/message/2366):

Johan, I can take on a user documentation role. Can't really help with coding, but I do write professionally: not strictly technical writing, but mostly user training materials in design and technology fields, or making business topics clear and readable for non-specialists. English-only, I'm afraid.

**Comment by objectivesea** [comment]
**Proper, well-organized documentation will be of great help for both new and existing users.** I see it as proceeding from Chapter 1 — guidance for installation on Windows, Mac and Linux/UNIX; to Chapter 2 — a sort of "Hello World" example that is very basic but can immediately demonstrate the program's utility in generating a simple melody score or tablature. Subsequent chapters should each introduce one new feature and provide a worked example, showing the source code to be entered and the expected output.

While the rest of the book is being drafted, the first few chapters could be made available as a PDF file, along with a method for accepting suggestions for improvement, perhaps by making an editable Word or LibreOffice file available for download.

Once the English documentation is tested and found to be correct, to the point that people can follow the instructions and replicate the project examples, I would suggest that we need to make use of an easy interface such as **Transifex** or **Wikitrans** for translating the documentation into other major languages, like French, Spanish, German, Dutch, Esperanto, etc. People who are fluent in a non-English language, either native speakers or who have professional competence in that language, as well as possessing professional competence in English as a second or third language, would be ideal candidates for supervising a particular translation into their strongest language.

The English version will remain the basic source document, but the global translations will not need to wait until the canonical version is complete. After a few chapters are tested and complete, however, translators would be able to start their work. We would just need to assign version numbers and keep them consistent across languages. File naming conventions should also be established that incorporate the basic two-letter abbreviation, like IT for Italian or ZH for Chinese. Maybe Alyn, who modestly claimed only English skills, might also be able to work on the Welsh (CY) version when the time is ripe for that task.

I wish Alyn well, and I commend him for the most generous offer to facilitate the important ChordPro documentation project.

Kind regards,
Erik Bjørn Pedersen ("Objectivesea" at WIkipedia and Tatoeba)

**Comment by gwyndaf** [comment]
Just scratching a few initial thoughts, based on what I've had to explain to other people when trying to describe ChordPro, or help them use it.

* _Descriptive_ nature of the ChordPro format. Most people will likely be more familiar with modern wordprocessing, where there's no separation between content, structure/function and appearance. The idea of separating content/structure and styling remains a relatively modern design concept, so needs a little explaining. I've sometimes done that with a simple diagram that shows 'song' and 'styling' inputs to the software, and 'output'. Also explain merits of this approach (_e.g._ repurpose song content for different output and styling intents).
* Recognise that other software understands the ChordPro format. Not detailed, but simply enough to show how the format is intrinsically useful, because it's portable. This provides an opportunity to mention that the level of ChordPro support varies between software. It's outside scope to document compatibility level of other software, but maybe include/reference table of what features are supported by different releases of ChordPro. This potentially allows other software to more clearly identify its level of ChordPro support (_e.g._ v5).
* Language: the terms 'specification' and 'reference implementation' might be a bit technical or confusing, and terms like 'ChordPro format' and 'ChordPro software' might be clearer for general explanation.
* Software: it might make sense to give more/initial focus to the GUI, probably citing Windows examples, which is probably most common. However, this could potentially increase the user base reliant on the GUI, and thus create more demand for existing features to be available through the GUI, _e.g._ creating a songbook with `filelist` or clearer/simpler system font selection.
* Software focus or standard focus? Although the ChordPro format is the logical starting point, most of my conversations suggest that non-technical people tend to think of ChordPro as a software product. I suspect there might be a risk that people easily grasp the format, but then stumble when trying to install/run the software. So, for clarity, it might be more accessible to focus more on the software, and deal with the 'hard bit' first. That then lays the groundwork to then move progressively through simple to more advanced uses of the format and configuration/customisation.

Overall, I'd envisage something that's maybe more tutorial in nature, _i.e._ introducing concepts and practices roughly in the order they might be needed as a user progresses from basic to customised usage (rather than how they logically relate to ChordPro functionality). I'd see the core reader as someone who already understands musical concepts, is familiar with writing up songs on a wordprocessor, and is used to Windows or Mac based GUI software. Linux and CLI approaches still covered, but maybe not first.

Because 'tutorial' and 'reference' materials serve different purposes, this could exist alongside the existing documentation, which could continue to serve as a detailed reference, and be linked from tutorial. In the early stages, at least, that makes it possible to provide a clear grounding in using ChordPro, while still covering the full functionality in detail.

**Comment by gwyndaf** [comment]
@objectivesea Thanks for your thoughts, Erik. Although I am indeed Welsh, I grew up in England so my grasp of the language is very limited (but something I'm looking to improve).

I'd imagine that translation might be some way into the future but, for future-proofing, can you tell me if the translation interfaces you mention favour any particular document format?

Right now, I'm looking to choose a useful file format for outlining and drafting, which is clear and unobtrusive for writing, supports simple graphic illustrations and examples, is cross-platform and shareable for review/testing, provides a solid foundation for multiple output formats such as HTML and PDF with no additional reworking. And automated translation.

My background is more writing and communication than development, so I'm open to suggestions on a useful, well-established format. To get the work started, I'll most likely start with Markdown or LibreOffice Writer, which are more writing- than code-focused, and allow flexibility in repurposing the material later on.

**Comment by kh63894204** [comment]
Hi All,

I would like to offer my help with this documentation project in any way needed.  I have quite a bit of technical writing experience for both tech level and general end-users.  I'm located in New York and English is my primary language.  I am retired now so I have all sorts of free time to offer.

I've been using ChordPro for the past 7 months to build up a song library for our local ukulele club.  It was no small task for me to learn all the options and capabilities of this application.  I agree that creating guided tutorials would be more helpful for the typical user.

Ken H.

**Comment by gwyndaf** [comment]
Thanks very much Ken. I haven't yet given much thought to how the work could be divided up, and might wait a while to see who's interested/available to help and suitable workflow. However, some initial thoughts...

I've been using ChordPro for about 5+ years, and the development of my knowledge has largely tracked the evolution of ChordPro, rather than the perspective of a user who's new to what's now fairly mature software/standards. So, your more recent learning experience will be very useful, to both inform the content/structure and review for clarity. I've had some experience of trying to explain/coach ChordPro to new users, and insights into the experience of recent new users will help identify the main knowledge/understanding gaps that need filling. 

Right now, I wonder if you could perhaps reflect on the concepts or practicalities that were hardest to grasp when you were learning ChordPro, and maybe share those at some point, but no rush. For instance, you mention it was "no small task", so were there any particular obstacles that took most effort or tenacity to overcome?

**Comment by kh63894204** [comment]
Hi Alyn,

I was already well versed in the ChordPro format using another app for many years so I didn't have to learn any of that.  However, I wanted something that would give me much more flexibility in the formatted output so that's when I looked into ChordPro.

Getting started using the basic ChordPro program was really straight forward.  Even playing around with customizing the spacing, margins, fonts and sizes isn't too complicated.  Learning to do really fancy formatting with Pango was a learning curve for me (mixing fonts and colors on the same line, strike-outs, rising, etc...).  Also messing around with all the options in the config file to get the look that I liked took awhile.  I also spent a good amount of time modifying the ukulele chord library to my liking.  The wxchordpro.exe editor is also not feature-rich so it took me awhile to figure out a good work flow.

I think we could start out with making a top-level outline of the tutorial topics.  From there, we could divvy up the parts to assign to different people for authoring.

Ken H.

**Comment by gwyndaf** [comment]
Thanks Ken. Yes, I think identifying and prioritising the intended readership (and thus tone, focus and learning journey), and outlining a suitable sequence of topics for them will be important early steps.

Your Pango comment is interesting. I suspect that people more accustomed to wordprocessors might be tempted to ask "How do I make ChordPro behave like Word?", which is a valid question, but one I'd be keen to reframe rather than answer directly. That is, it's a fundamentally different approach, which would make more sense to someone familiar with early semantic HTML. Suggesting that it's anything like a wordprocessor could cause confusion later on.

Pango would be a more direct answer to the "like Word" question, such as directly styling things like typeface, size, colour _etc._ However, that immediate gratification might be at the cost of fully understanding ChordPro's 'primary' features like directives and configuration. That is, it's not an efficient use of ChordPro to, say, put a whole chorus in bold by using Pango.

I'm thinking that the hierarchy/sequence in which to present ChordPro functionality might roughly reflect its evolution and refinement, _e.g._ from line-based to character-based formatting. My instinct is that Pango should be introduced later, and would then be useful groundwork for configuration options and designing a custom layout.

Another consideration (although it might conflict a little with the above) is to focus initially on what can be done through the GUI. 'Functional' features like transposition might be a bit more advanced than basic layout and formatting, but they also represent a good reason for favouring ChordPro over a wordprocessor, so covering them early would give even basic users some 'quick wins' for choosing ChordPro.

Incidentally, as you play ukulele, I've just updated my suite of ukulele chord definitions (C, G and D tunings), and the master file from which they're generated, if you'll find those useful: https://github.com/ChordPro/chordpro/issues/220#issuecomment-2221155191

**Comment by sciurius** [comment]
It is great to see you guys working together!

I'll try to stay out of the loop as much of possible, but there are some points that I want to address at this stage.

I'm a strong believer in version management. The documentation should therefore be written in, or can be converted --lossless!-- to, a textual format (e.g. markdown) that is under revision management (e.g. git).
Also, it should be possible to produce a (section of a) web site from the textual format.

The current documentation is written in markdown and processed with hugo to create static web content. This doesn't need to stay that way if you have better alternatives.

Initial focus on the GUI is a good idea. You may want to use the (new) macOS GUI since that is much more powerful than the classic GUI.

As for chordpro markup (I no longer call it pango): this can be integrated in the GUI. You create bold and italic text and when saved this is turned into markup.

A big thank you to all of you!

**Comment by kh63894204** [comment]
Alyn and Johan,

I'm in agreement with your comments.  I probably shouldn't have said that it took me awhile to learn ChordPro since the Pango styling stuff is not technically part of ChordPro.  However, for me it was what I specifically wanted the app to do.

Once a person gets past the basic ChordPro configuration and usage, I think that output formatting and styling is the next thing that people will tackle.  I see a fair amount of questions on the forums relating to this and that's where I usually got the answer to my own newbie questions.  The (excellent) ChordPro reference documentation does contain all the necessary information but it's not in an easy to follow guided format.

In my opinion, I think that nowadays most people would like ChordPro to be like a word processor.  I'm hoping that the Windows GUI (or a web front end) will receive some more development love.  I usually don't use a Mac but I am very tempted load up a VM to try out that new GUI that was recently developed.  A great GUI would have a big impact on the ChordPro user experience.

Thank you for the ukulele chord libraries.  I will check them out.  I haven't uploaded my chord diagram changes yet because I'm still checking all of my chord sheets through them.

Ken H.

**Comment by sciurius** [comment]
> most people would like ChordPro to be like a word processor

True, as long as the underlying format allows being able to exchange the songs with just about everyone else.

> I'm hoping that the Windows GUI (or a web front end) will receive some more development love.

Yes, it would be nice if someone stepped up, just like Nick did with the macOS GUI.

I do have a prototype chordpro server that runs in a docker and communicates via a basic web frontend. It quickly runs into problems since the server runs in a different environment as the user at the frontend (think config files, fonts, images, ...).

A number of years ago I attempted the HTML + CSS → PDF route but there were (and still are, AFAIK) no tools that can produce a decent PDF from HTML + CSS. 

FYI.
https://chordpro.org/dl/ChordPro-6.050_081.dmg
![Screenshot 2024-07-11 at 09 24 27](https://github.com/ChordPro/chordpro/assets/1118804/6e464202-51d6-4c41-a4d0-29b73b81960c)

**Comment by gwyndaf** [comment]
Ken, I agree that formatting will be a key aim for most users, and the output can indeed look similar to a wordprocessor's. However, I think it would be a mistake to suggest that ChordPro _functions_ like a WP, even if future GUI developments move it more in that direction: right now, the aim is to provide guidance for using ChordPro as it currently is. I'd like to equip the user not only to get the results they want (or understand why not), but also to do that with least effort for a song; otherwise they might as well use a WP.

I do believe that helping users grasp the conceptual model of any software is an important step in providing a solid knowledge foundation. For instance, I often have a conversation with designers, who conceive a design and approach in terms of how InDesign functions (page-by-page layout model); but they believe it can be easily translated into MS Word (continuous text flow model). That it, they assume both applications follow the same model, but they fundamentally don't, even though the ultimate result of both is to organise ink on a page (or PDF).

So, I certainly agree that it'll be useful to recognise that users might want similar _output_ results to a wordprocessor, and that formatting will be a high priority for many. But, to help them get there, I think we need to clearly establish the (different) working model of ChordPro.

Although inexperienced WP users might simply dive in and format text manually, experienced users should already use and understand styles (Heading 1, Body Text _etc_), so they should readily grasp the ChordPro concept that we're describing the _function_ of different elements, not simply their appearance, so that could be a useful foundation to build on - certainly as an underlying concept, even if the mechanics differ.

There's potentially a tension here between 'tutorial', 'reference' and 'FAQ' approaches. That is, the clearest order for introducing basic concepts and methods to a new user might not be the same as the logical organisation of reference material for an experienced user, or the task-oriented 'how do I...?' of an FAQ. All are valid and useful approaches, but I think the main need right now is to provide a way in (tutorial) for new users.

So, although formatting is one big topic from a wordprocessor perspective, I think it'll need introducing in progressive (and possibly non-contiguous) stages that reflect different aspects of ChordPro. For instance, some initial thoughts based on _current_ ChordPro functioning and features:

* Song elements and structure (essential foundation)
* Formatting via GUI (selecting a 'template' config, although I haven't seen the new MacOS GUI options)
* Formatting via directives (line-based)
* Formatting via Pango (character-based)
* Formatting/layout via custom configuration (probably focus on PRP rather than JSON format)

Right now, we have the challenge that some tasks (like setting page margins) that are standard starting points for a WP user, require configuration editing in ChordPro, which I'd consider a more advanced topic, and quite alien for non-technical users more familiar with GUI wordprocessors.

One way around this, to provide quick wins for new users, could be to offer a more extensive gallery of preset layout designs (config files). That's not strictly documentation, but enhances usability by reducing the need for manual design/layout configuration.

**Comment by sciurius** [comment]
I agree that setting the mental model has priority. A ChordPro song is a text document (which may be stored as a file on disk, or on the net, or in a database, etc.) that is maintained using a text document editor. The ChordPro tool (embedded, or explicit) creates a formatted song document (PDF, which may be a file on disk, etc.).
I recently realised that modern computer (ohone, tablet) users look down on text editors but have no problem deploying those tiny widgets in mail programs and social media tools -- which are no more (often less) than very simple text editors.

I think following "Formatting via GUI" there should be a section "Using the CLI".

Please avoid the term "pango". While the Pango Markup Language was an initial inspiration for the ChordPro in-line markup they have diversed in very incompatible ways.

The preferred config format is Really Relaxed JSON (RRJSON), which is more similar to PRP than to strict JSON. There's a tool to convert JSON config to RRJSON (it even adds comments!).

Writing documentation will inevitably lead to software changes (e.g. to resolve inconsistencies) so be prepared to run the development version from git.

**Comment by gwyndaf** [comment]
From a user guidance point of view, I'm inclined to adopt the perspective that text editors, plain text and CLI instructions might be unfamiliar and intimidating to people who've grown up using GUIs and more interactive tools. Not so much 'looking down' but more pragmatically recognising that there's a potential user base who don't even engage in the forum: I've noticed a few people leaving when it seems like discussion might have turned too technical for them.

On the hand, we also need to equip users to exploit the full power of ChordPro, which will need to tackle CLI and configuration at some point. I think my aim is to lead readers there in a manner that means they're sufficiently invested by then and don't get deterred at that point. That'll need a bit of thought to chart the most engaging and accessible journey for the target user.

Points noted about avoiding the term "Pango", preferring RRJSON config format (which I'll need to explore myself), and working to evolving git versions.

I think my next step is to take a little time out from discussion to do the preparatory conceptual work, like defining target audience, purpose and aims of documentation, and thus outline/ordering of topics, for which current discussion has been useful. It'll probably be more productive to write those up in a well-organised fashion, so there's something tangible around which to focus further discussion.

**Comment by kh63894204** [comment]
Alyn,

I think the audience, intent, and approach are important decisions that need to be made before diving into content production.  I'm glad we're giving this a lot of thought.

You asked me what my experience was like as a new ChordPro user so I was just rattling off my thoughts.  I am a fairly technical person so I'm probably not a typical new user.  I do have a lot of experience simplifying subjects from a neophyte perspective.  I've actually found it a little difficult to explain to someone why they would use ChordPro instead of a word processor.  I'd like to entertain you with how I see a typical new ChordPro user:

Let's start by assuming that a person is interested in making a chord sheet.  In all probability, they would simply jot something up using a word processor.  As they progressively get more sophisticated, they want to add chord diagrams.  That's when they would go searching the web for something to use.  Let's say they find a website that generates chord diagrams and they cut-n-paste those into their word processor.  All is good in the world.

As their experience making chord sheets progresses, they find adding chord diagrams to be a major pain.  They also realize that sharing their chord sheet in a proprietary format is not ideal.  Time for another web search.  For the sake of brevity, let's just say they discover the ChordPro website.

I think a person reading the ChordPro home page doesn't really see any compelling reason for using the program.  However, it's free so they download, install, and run it.  Now what?  This is the point at which you'll either lose someone quickly or they decide to try it out a bit.  If they chose the later, they will quickly encounter a lot of unfamiliar techno jargon and a much different work flow from a word processor.

I believe that this is the point at which we are looking to provide some help.

Ken.

**Comment by kh63894204** [comment]
Johan,

Please excuse my use of the term "Pango".  I do understand that it should be properly called the "ChordPro markup language".

Ken.

**Comment by sciurius** [comment]
Hi Ken,

No need to apologise, it was not a complaint. It is just that I want to tackle these misnomers as early as possible.
And, not being a native english speaker, my remarks sometimes sound stronger than necessary. Sorry for that.

Finding the right descriptive terms for some ChordPro concepts is crucial to make sure we are all talking about the same things and to provide unambiguous information to the users.

For example, the term `ChordPro` is sometimes used to designate the file format, and sometimes to designate the program. I think `ChordPro markup` better covers the typical ChordPro directives, so we need another term for the in-line markup. `Printing` is used quite often, although nowadays the PDF documents are mostly viewed on screen instead of printed. Chord `diagrams` versus chord `grids`.
A bigger problem is `font`, that is used for several font-related topics.

I hope we can work out a good set of terms to solve these ambiguities.

**Comment by sciurius** [comment]
> Let's start by assuming that a person is interested in making a chord sheet. In all probability, they would simply jot something up using a word processor.

I think most ChordPro users start with a ChordPro document they got from some source.

**Comment by objectivesea** [comment]
Hi, Alyn,

I may have misled you about Transifex. I had thought that they had a free level for non-profit or open-source developers, but it looks like their minimum charge is now $120 a month, which seems excessive, even confiscatory, to me. (https://www.transifex.com/pricing/). And Wikitrans is an app specifically for translating the English-language Wikipedia to Danish and Esperanto, so my memory was off on that as well.

Transifex’s interface, as far as I remember it, divides material to be translated into individual sentences, and makes use of a translation memory so that similar phrases get translated in a similar and consistent way. That can be convenient, especially for interface translation, but the lack of this feature should not be an obstacle when tackling documentation. 

(By the way, the file format is a «text-based portable object file («.PO»).There’s some information here:  https://www.simultrans.com/blog/what-are-.po-files , and there are likely some freeware or shareware alternatives to Transifex.)

If we have a document, perhaps in LibreOffice Writer, that makes use of defined styles and applies them rigorously, a prospective translator can work in place and make sure the correct style is still selected. Or a translator who works in another word processor can paste translated text over the original text, making sure that pasted text is unformatted; then if a few words in a paragraph need to be bold or italics, these styles can be reapplied selectively and as needed after pasting in the translation.

If you choose to use Markdown, I think similar considerations might apply; the syntax seems simple, and the translator should just make sure that the formatting of the translated text matches the original English version.

If you devise something like a sample page or two of proposed documentation, I would be happy to work up an Esperanto version as a test of the translation process, so that we can see how smoothly the process goes and to suggest work-arounds if anything does not go smoothly for some reason.

Once the ChordPro application reaches a plateau where it’s relatively feature-complete, you’re happy with the GUI, things are running very smoothly, and you foresee no major changes for the nonce, it may be time to begin internationalizing the application by producing localized interfaces. This is the kind of task that programs like Transifex (and, hopefully, freeware alternatives) are ideal for.

Kind regards,
Erik

**Comment by gwyndaf** [comment]
No problem, Erik. I don't see translation as a high priority at this early stage, and not something I have a background in. ChordPro format and software are English-based, so we can't (currently) offer users a wholly native-language experience. However, it's a useful consideration, which I'll incorporate into 'big picture' thinking and outline planning.

Right now, I think it's most useful simply to adopt a flexible working format that will be easily shareable and doesn't rule out translation or other repurposing at a later stage. Markdown seems the simplest option, certainly for structured text, probably with GraphViz or Mermaid for simple conceptual diagrams, SVG for more complex vector illustrations and PNG for raster graphic examples.

Once draft content in that format starts evolving and being shared (probably github for simplicity), I hope that should make it easy for you to grab a sample and test out its suitability for translation.

**Comment by Desbeers** [comment]
ChordPro, the application, is an editor. It is also a reference implementation and the first impression. So, what to show when you are opening a ChordPro file?

Do you expect plain text? Do you expect a PDF render of your song? Or side by side?

I wrote my own program long time ago to show me my songs. Rendered from ‘ChordPro’ markup. Chordpro is the musical version of markdown for me. Plain text, simple, readable as-is but made as beautiful as you want it to be.

So, if it is up to me, no ‘markup’ buttons in the editor. ChordPro is Markdown. Not a weakness; a strength!

In my -very opinionated opinion-, I do not like the first impression of the generated PDF when opening my first ‘chordpro’ file in the GUI. Most probably you found it somewhere and expecting something nice. I see an editor, can open a preview and see something I don’t like too much.
 
Well, I wrote the new macOS wrapper. Better, but for sure not great.

In my -again- opinion, documentation is nice for those who read it but the application sets the first impression. That is my focus.

When I open a ChordPro file I want to play a song.

**Comment by gwyndaf** [comment]
Thanks Nick. One of my main aims (as a writer/educator) is to equip a less tech-savvy user to feel more comfortable with using ChordPro _as it currently is_. I'm not a developer so  - aside from raising issues and suggestions - there's not much I can do to advance the software's progress! 

In terms of the overall user experience, ease of use will likely come from a blend of GUI design and guidance, although they may come to inform each other. Of course, the ideal is that the whole process becomes so intuitive that no guidance is needed, but I doubt we'll reach that point even with a lot of development effort for the GUI (my Mac usage is sporadic, so I haven't seen your wrapper yet, I'm afraid).

For example, the ChordPro format, of putting instructions within the document, will be a new and alien concept to users who've only ever known WYSIWYG wordprocessor and DTP software. Not hard to grasp, but does need explaining, because it's quite fundamental.

Another example: the core concept of separating content (ChordPro source) and styling (configs) is fairly specialised and will be a bit abstract for many. It'll be familiar to users with HTML/CSS experience, but not many others. I've known many users (with and without design backgrounds) who only think in terms of putting 'ink on the page'. I've even worked with professional designers who haven't grasped the idea or usefulness of styles (or style sheets).

Even without procedural guidance, there's some conceptual background that needs explaining to help users get oriented for using ChordPro, rather than get frustrated and put off using it. So, this work will complement any advances in the GUI that might come with time.

For now, I'm happy to write guidance for users, and won't be offended if it turns out that no-one needs to read it!

**Comment by Desbeers** [comment]
Thanks gwyndaf for your reply. I gave a lot of thought.

First of all, I’m a *user* of **ChordPro**, my format for many years, so, the format is obvious to me and sometimes I forgot I had to learn it as well.

Second, my wrapper looks nice, however, its only *on pair* now with the Linux and Windows wrapper. It does highlighting of the text as extra and makes its easier to preview, but fundamentally, it is the same.

That was my goal.

Now that it’s done, I don’t like the result.

I never really used the official **ChordPro** program with all its possibilities.

Playing guitar is what we want. We don’t want to fiddle with configuration files, strange markup, JSON files; we want to play the guitar and have a great songbook to help us!

So, besides good documentation, the basics should give you a pleasure to use **ChordPro** to play your songs.

Opening an existing **ChordPro** file should give you a beautiful rendition of the source that you can’t wait to play.

So, it should not show you a text editor. It should show you a song.

When you start a new song, show a guidance. When you edit a song, show help when you open the editor. The current documentation is written in markdown, and same as the **ChordPro** format, it can be shown however you like! On the website, obviously, but also inside the application.

I’m fiddling around with my macOS wrapper. My dev-version opens existing files just as a PDF preview with a button to edit it. A new song shows a welcome screen.

<img width="1012" alt="Screenshot 2024-08-06 at 15 14 02" src="https://github.com/user-attachments/assets/d402e2f0-2196-41e7-8456-8120f41a0db1">

From a GUI *programming* point of view, it’s easy for me. From a *user* point of view; there is a lot to learn.

I want to make it a great **ChordPro** viewer, simple to use and easy for small modifications, but offer the advanced possibilities for those who want to tailor the resulting PDF to their own needs.

> If it is up to me, the GUI will never be a WYSIWYG editor. That will abstract the strength of a plain-text format into just another implementation. ChordPro is plain text and you can use and read it however you like. Again, it is musical markdown. There are many markdown editors but the best ones are the ones who does not hide the source but guide you.

<img width="570" alt="Screenshot 2024-08-06 at 15 49 29" src="https://github.com/user-attachments/assets/614b95f4-6ed1-4b35-96db-d8a0b03f6643">

**Comment by Desbeers** [comment]
I want to add that the GUI for macOS, Linux and Windows should be reasonably on-pair. I can’t make the macOS wrapper super-fancy and leave the other platforms behind.

The GUI version will most probably give the first impression but the strength is at the command-line. Also something to think about...

**Comment by gwyndaf** [comment]
Thanks for your comments @Desbeers. As previously, these seem to relate more to the software and UI design, but this topic is about developing new documentation.

I totally agree that the _overall_ user experience can be enhanced, and the need for documentation reduced, with product/UI design enhancements, so they are related. However, I'm an educator/facilitator/writer, not a software developer, so there's not much I can do about software development within the scope of the documentation work.

From a project management point of view, it's theoretically possible to wait for an enhanced GUI before starting on documentation, but my thinking is to start work based on the _current_ UI. If and when they get enhanced, that should mean only a reduction in the guidance needed.

Perhaps you'd like to open a new issue for discussion of UI enhancements that might make ChordPro more easy and enjoyable to use?

I'll be happy to contribute thoughts to that topic. I'll need to use the GUI approach quite systematically (and with a novice's eye) for tutorial purposes, so might notice aspects that seem to need a lot of explanation, but could be addressed with small GUI changes.

**Comment by Desbeers** [comment]
Sorry for ‘highjacking` the subject; did it with my best intensions.

I will just follow this discussion and if I can do anything related to the GUI just let me know!

**Comment by Desbeers** [comment]
I don’t talk so fancy; I submit pull requests.

**Comment by gwyndaf** [comment]
No worries. I can't code anything that's worth pulling, so this is what I can do!

Thanks for the offer regarding the GUI. I'm sure thoughts will occur to me once I start thinking about how a novice user might try to approach ChordPro.

**Comment by sciurius** [comment]
The [beta site](https://www.chordpro.org/beta/) has a new search facility. Please check it out.
The old search was dysfunctional for a long time, apparently noone missed it :).

**Comment by gwyndaf** [comment]
Great news. I tried the old search once and found it non-functional, but pages (including beta) seem fairly well indexed on Google, so I've tended to search via that route.

A thought: does the new search log the search terms? If so, that data could provide useful insights into which topics users need most help with, which could help focus user guidance.

**Comment by gwyndaf** [comment]
I found the search box at the top of the navigation sidebar on the beta's implicit home page `https://www.chordpro.org/beta/`, and it seems to work nicely.

However, the search box (or any link to it) seems to be missing from navigation bars on other pages, including the _explicit_ home page `https://www.chordpro.org/beta/home/`, and there's nothing on the navbar that directs back to the home page with the search box.

As a result, someone who's already reading part of the documentation doesn't seem to have any direct route to the search feature. I'm guessing it just needs the navbar _with_ the search box to be replicated to other pages on the site.

**Comment by sciurius** [comment]
With 6.060 comes a new landing page and an improved search facility. Please let me know what you think.

**Comment by gwyndaf** [comment]
Landing page is looking great!

Some general thoughts:

1) Maybe add a simple 'Documentation' link to top right menu bar. Although the 'Getting Started' button leads there, that might be counter-intuitive for users who have already gotten started and need to check some usage syntax etc.
2) It'll be useful to consider positioning and target audience, which might determine how the information's organised. For instance, if the aim is to attract musicians who aren't necessarily technologists, tunings and diagrams might be a higher priority than, say, Unicode support. I've been thinking about something like a "Why ChordPro?" section in my initial outline for user documentation, so further thoughts will follow from that.
3) On a small/low-res laptop screen, the nice graphic occupies most of the screen, so even "Welcome to ChordPro" is nearly off the bottom of the screen. On a mobile phone, it fits perfectly, with the 'Get Started' button nicely at the bottom of the screen. _But_ that might have the effect that people just click the button, and don't scroll through the remainder.

Just initial thoughts, and generally looking great, though.

Seems to be an issue with the Search result links leading to missing pages (404). I think it's this:

* {start_of_chorus} page URL is `https://www.chordpro.org/chordpro/directives-env_chorus/`
* Search result link is `https://www.chordpro.org/directives-env_chorus/`

So, it's maybe that the search links are pointing to the right page, but at the wrong level.

**Comment by sciurius** [comment]
> Maybe add a simple 'Documentation' link to top right menu bar.

Done.

> On a small/low-res laptop screen, the nice graphic occupies most of the screen, so even "Welcome to ChordPro" is nearly off the bottom of the screen

What resolution is that?

> Seems to be an issue with the Search result links leading to missing pages

Fixed, at least for the time being. As a side result, beta site searches now also go to the stable pages. I need a better fix for that.

> It'll be useful to consider positioning and target audience, which might determine how the information's organised. 

Your thoughts are very welcome!

**Comment by gwyndaf** [comment]
> What resolution is that?

1280x800, though I'll admit it's an old laptop.

Search results links working nicely now. I see I can still get straight to beta docs at `https://www.chordpro.org/beta/` and will just avoid searching those, which hasn't become my standard practice yet!

**Comment by sciurius** [comment]
Searching in the beta docs is now also fixed.

**Comment by Desbeers** [comment]
So, all the fancy talk in this issue leads to no result so far.

Meanwhile, the **ChordPro** GUI is completely rewritten. The documentation website looks better.

Put your contributions where your mouth is. 

I did.

Do you?

**Comment by gwyndaf** [comment]
I know a new GUI is in development, but don't believe it's part of the the official release version yet, so I haven't seen it.

The GUI is central to my thinking/planning for the tutorial, because the command line is a big obstacle for non-technical users, so I want to ease new users into ChordPro via the GUI, and I'm looking forward to seeing the new one.

Until then, there's no sense wasting time and effort writing a tutorial (with screenshots etc.) based on the current/old GUI, if it's going to be radically changed in the near future.

**Comment by sciurius** [comment]
The new GUI may be considered complete but it requires some updated system libraries which makes it tricky to run/install from source (CPAN) or git. I think most (GUI) users will use the prebuilt packages anyway.

I'll make pre-release packages for Windows and macOS, and a docker image for you to play with.

**Comment by AnniePerson** [comment]
I'm also interested in a pre-release package for windows of the new GUI. As I have no Mac I cannot take a look at the macOS version.

**Comment by sciurius** [comment]
Package for [Windows](https://chordpro.org/dl/ChordPro-Installer-6-60-176-1-msw-x64.exe) and [macOS(intel)](https://chordpro.org/dl/ChordPro-6.060_176-macOS-INTEL.dmg).

Docker image is on the Docker HUB, chordpro/chordpro:v6.060.176 .

@Desbeers Do we need an ARM image?

**Comment by Desbeers** [comment]
Why not? The more testing the better! I've sent you an email.

**Comment by AnniePerson** [comment]
Thank you

**Comment by sciurius** [comment]
> The GUI is central to my thinking/planning for the tutorial

Indeed. But there is more. After some recent discussions I think this can be a first step to an outline:

* Introduction
* Using the ChordPro GUI (extensive)
* Using the ChordPro CLI (concise)
* Basic ChordPro (elementary directives and chords, transposition, no fancy stuff)
* Advanced ChordPro (complex directives with attributes, transcoding, images, delegates, config files)
* Showcases/portfolio
* Appendix: Configuration file reference
* Appendix: Installing ChordPro

**Comment by sciurius** [comment]
(macOS ARM)[https://chordpro.org/dl/ChordPro-6.060_176-macOS-ARM.dmg). Thanks, Mick!

**Comment by AnniePerson** [comment]
The outline of the docu looks reasonable. Maybe it should mention where chord grid, ABC and Lilypond should be included. I would expect ABC and Lilypond in "Advanced ChordPro". Not sure if chord grids should be in "Advanced ChordPro" or "Basic ChordPro" with some tendency to "Advanced ChordPro"

**Comment by gwyndaf** [comment]
> a first step to an outline:

Thanks Johan. My initial thinking is that we've probably got three broad categories of need:

**Getting Started tutorial (user-centred)**
This is my main focus, partly because I'm experienced in both writing and training, but also because I've tried (and often failed) to encourage non-technical users to use ChordPro, so I've a reasonable idea of what some common obstacles are. Those include technical competencies like using the command line, but also conceptual blocks like the separation of information and styling, which is still alien to people used to WYSIWYG wordprocessors and page layout software.

An important intent is to attract and engage new users, especially if they're put off by the command line. So, my broad aim is to guide the technically-novice (but musically-literate) reader on a journey from understanding ChordPro's conceptual background, through using the GUI to carry out various basic tasks, then (once they're sufficiently engaged to want to dig deeper) into configuration and use of the command line for more powerful usage. How I chart that journey will reflect the GUI, which is about to change.

I don't envisage the tutorial being comprehensive in its coverage, but at least give the novice user the conceptual and practical grounding to then make use of other areas of documentation, for which I see two further areas...

**Cookbook (task-centred)**
Answers to common "How do I?" questions, such as "How do I change the title font?", which often arise from users more familiar with wordprocessors or layout software. The user forum could be a good source of such questions.

Initially, it could simply be a list of common tasks/questions, and those might naturally suggest a grouping/structure as they evolve, e.g. 'song formatting' or 'working with multiple instruments'.

Target audience is those who've worked through the tutorial, so have a grasp of basic concepts, but maybe still need some help putting that knowledge into practice. Each task example could also link to...

**Reference (software-centred)**
I think current documentation does this job pretty well, in providing an authoritative and detailed account of every aspect of ChordPro's functionality. The main challenge is probably getting users to a point where they can make effective use of this information, which I think the Tutorial and Cookbook areas should help achieve.

There might also be a little top-level restructuring needed to align those three areas, and also include some sort of 'shop-front' for curious potential users, including the showcase.

**Next steps**
For me, I think my priority will be the tutorial, which I've already been putting time into conceiving. Thanks for the pre-release GUI versions, which will be an important factor in refining the conception and translating it into substance. Once I've got something tangible on that aspect, I'll probably open it as a separate discussion topic, with a more specific focus.

If somebody wants to take ownership of the 'cookbook' section, I'm happy with that, otherwise will move on to that myself once the tutorial's done.

**Comment by sciurius** [comment]
> Maybe it should mention where chord grid, ABC and Lilypond should be included.

Yes, as you already assumes this is (more or less) advanced stuff.

My aim at 'basic' is the vast majority of ChordPro songs that you can find on the net. ChordPro 3.6 (see the [cheat sheet](https://www.chordpro.org/beta/chordpro-cheat_sheet/)). Maybe some of the newer features provided they fit into the simple `{directive}` or `{directive: arguments}` form.

**Comment by sciurius** [comment]
[This is following up from thread #373]

> I know we talked about this before, the default extension for a ChordPro document. It is 2024; there is simply no (technical) reason to not default to .chordpro anymore.
> [Wikipedia](https://en.wikipedia.org/wiki/List_of_filename_extensions) does not make standards. The ChordPro Team does; we decide.

If I'd consider changing the default, I would first ask the users what their preferences are.

But ok, let's head for the 'Christmas 2004' release!

* Core: There is one experimental feature (the chord changes) that is partly in the core but not functional. We can release without.
* GUI: I'll address the issues you reported and finish the electrickery.
   - Chordname upcasing when typing a closing `]`
   - Expanding an abbreviated directive into 'full' form
* Maybe a couple of bug fixes but nothing that should impact documentation.

> Replace the ‘landing page’ with mine

Ok. I'll move the examples to a showcase.
@gwyndaf Would that fit in a 'cookbook' or should it be separate?

> old stuff should be removed or re-written (is done now I think)

I'm not sure what you mean? The install/getting started should be uptodate now but the rest of the documentation needs updating. In particular, all examples of JSON config should be changed into the new RRJSON format. 

> Announce the completely rewritten GUI on the mailing list? As beta and make builds publicly available?

Good plan. Did you complete the GUI docs yet? At least in such a form that it can be used by testing users?

**Comment by gwyndaf** [comment]
I think the current 'showcase' examples are probably best left as part of the 'shop window' function of the website, to show prospective users _what can be done_ with the ChordPro software, rather than explaining _how to do it_.

I'd envisage the cookbook as isolating specific tasks/outcomes (_e.g._ change heading font), and explaining how each can be achieved, whereas the whole-page showcase example combine several different techniques.

At some point, there might be a useful tutorial section of a complete 'worked example' that takes one output sample from the showcase and works through the various techniques used to achieve that, but I don't see that happening for every showcase example.

Let me see what I can do regarding GUI documentation in the next week or two. My main overall challenge is working out how to guide non-technical / non-designer users through the underlying concepts and mindset, but the GUI aspect should be more practical than conceptual, so probably a bit easier/quicker.

For screenshots, I'll work from pre-release version you [shared previously](https://github.com/ChordPro/chordpro/issues/390#issuecomment-2516320297), for Windows, which I guess most non-technical users have. Maybe you have download stats to confirm that?

**Comment by Desbeers** [comment]
> > old stuff should be removed or re-written (is done now I think)
> I'm not sure what you mean?

I was just talking about the GUI part of the documentation. ‘Getting started’ is done (but can be improved; no ‘songbook’ mentioned for example), ‘Creating a config’ has some issues (see the other treat) and there are still some ‘Preferences’ instead of ‘Settings’ on several places. I’ll take care of that.

> In particular, all examples of JSON config should be changed into the new RRJSON format.

Ah! Good one! I’m not so familiar with that format but I will give it a look.

**Comment by sciurius** [comment]
> Ah! Good one! I’m not so familiar with that format but I will give it a look.

There's a script `rrjson.pl` in the `script` folder that can convert JSON to RRJSON.

**Comment by sciurius** [comment]
> I think the current 'showcase' examples are probably best left as part of the 'shop window' function of the website,

Nick's proposal is to replace the 'shop window' by a [basic landing page](https://chordpro.org/landing.html), partly for simplicity, but also for technical reasons (the current landing page is alien to the rest of the site). Maybe it is better to replace the top part of the current landing page by the new content, and leave the rest until we have something better? The current landing page is already basic unless you scroll down.

**Comment by sciurius** [comment]
I'm aiming at a feature freeze tonight. Then we can make definitive screenshots and beta kits.

**Comment by AnniePerson** [comment]
Just started testing 6.60.176. 
As soon as a song contains an ABC section i get "Error: SVG Parser: No elements at C:/Program Files/ChordPro.ORG/ChordPro/script/../lib/SVGPDF.pm line 282."
![AfterYouveGone_ChordPro_6 60 176](https://github.com/user-attachments/assets/b49b43a9-df71-4c99-8e57-2431e87f43c1)
[AfterYouveGone_ChordPro_6.60.176_err.txt](https://github.com/user-attachments/files/18062925/AfterYouveGone_ChordPro_6.60.176_err.txt)

**Comment by Desbeers** [comment]
> Nick's proposal is to replace the 'shop window' by a basic landing page, partly for simplicity, but also for technical reasons

I looked (briefly) into the source of the current landing page; it’s a *beast*, I’m not going to touch it :-)

I will add a bit more ‘spice’ to my proposed page...

**Comment by sciurius** [comment]
> I’m not going to touch it :-)

NP, I will.

**Comment by sciurius** [comment]
@AnniePerson Great reporting! I can reproduce this with the latest versions. Working on it!

**Comment by sciurius** [comment]
For the documentalists: The ChordPro editor has some 'electric' actions.
1. Typing a `]` will look for a preceding chord name, and upcase it. E.g. `[` `b` `7` `]` → `[B7]`.
2. Typing a `:`, space or `}` after what looks like the beginning of an abbreviated directive will expand it. E.g. `{` `c` space → `{comment: ` .
3. Enter before a `}` will insert **after** the brace. Try Alt-I `k` `G` Enter.

**Comment by sciurius** [comment]
Ok, I think I have most issues solved. New klts to test:
https://chordpro.org/dl/ChordPro-6.060_194-macOS-INTEL.dmg
https://chordpro.org/dl/ChordPro-Installer-6-60-194-1-msw-x64.exe

If there are no apparent problems I'm going to announce these on the forum.

**Comment by gwyndaf** [comment]
Just reviewing the new GUI alongside conceiving the learning journey for novice users (rather than strictly testing). A few thoughts:

**1) Configuration editing.** Based on forum queries, I think the main interest, beyond basics, will be to customise output with config files. So it would be nice if the GUI editor could also handle config files - of course it can do it currently, but user experience is less than for song editing. Some suggestions:

* In the File > Open dialog, add an entry in the 'file type' drop-down for ChordPro Configuration files (.json, .rrjson), but still default to ChordPro songs.
* Add syntax highlighting for JSON, if a config file is opened.
* Potentially, there could be a 'Configuration' menu entry, but I think that risks confusion with 'Settings'. That is, most users will have seen settings/options/configuration words all used in menus to essentially mean the same thing. For now, probably best to keep config editing as 'second level' knowledge, but help enterprising novices by making it clearer (as above).

My thinking is that config editing will have more (new) user demand than CLI usage, because many CLI operations (transpose, transcode, config selection) are now available via the GUI. CLI also represents a _huge_ learning process for (prospective) users who've only ever used a GUI - not just starting a terminal, but also using/knowing the relevant commands, and how to navigate files and folders via CLI. And I don't want to put people off ChordPro too early in their usage, if they think CLI is essential. It's easy to overlook that for expert users/devs who regularly use CLI. I will include a CLI section, but realistically that will have to assume users are already familiar with using terminal/CLI, which will make it one of the most advanced topics.

**2) Songbook Builder** Really useful addition!

Suggestion: could/should this feature be added to the 'Tasks' menu, so it's accessible from the standard editor interface, without needing to go back via the startup wizard?

One testing issue (Linux 6.060_176): 'Add a standard cover page' with text for title and subtitle doesn't always add a standard cover page.

If (in Settings > Presets) I use defaults, including my core custom config, I get no cover page. If I ignore defaults and chose, say, 'Modern 1' preset, I get a nice cover page, as expected.

But, my own config's quite complicated, and I'm sure that's the cause, but Messages doesn't give me information to narrow it down. This might not be a major problem, though. Given the high likelihood it's caused by my config, it might never arise for basic usage with just built-in presets or simple custom configs.

**3) Startup screen / wizard**

Probably handy for total beginners, and becoming common on other software, but maybe adds friction for experienced GUI users wanting to get straight to work, so some thoughts:

* Maybe provide an option not to show the startup screen (go straight to editor interface). This seems common on other software (e.g. Adobe Creative Suite) that recently introduced a 'startup screen' approach.
* Incorporate into regular menu any startup screen operations that aren't already there - main one that occurs to me is the Songbook Builder, which could go under 'Tasks'
* It's maybe a bit counter-intuitive to get back to the startup wizard by using 'File > New', so maybe clearer if there's an explicit menu entry to call up the startup screen / wizard again.

**4) Minor thoughts**

Maybe we don't need 'Insert > Key', which could become a rabbit-hole of what other metadata to provide via the menu. Title, Subtitle, Artist seem to make sense because (I think) they're included in headers for the built-in layout configs, as a coherent basic-level experience. So it's maybe sensible if the 'Insert' menu metadata items correlate with what metadata is output with the built-in layouts.

If we give users the option to insert a key, which doesn't then appear in built-in PDF output, that could be confusing. NB this is very much 'basic level' usage, and next-level users who can create custom configs should by then know how to add other standard metadata manually, in the editor.

I think these are all potential refinements to simplify/clarify the user journey, rather than functional issues, so no reason they should hold up release.

**Comment by sciurius** [comment]
Great feedback! Excellent!

Configuration editing is something that we definitely need. As it is structured internally it is rather straightforward to add a config editor but...
Config files (at least the modern ones) are relaxed JSON, and there is no highlighting available. We would need to develop a custom lexer (as Scintilla calls it). I have plans to do so, for rrjson and for chordpro, but it will take time. 
Raw editing (w/o highlighting and electrickery) would be feasable.

Cover page: this makes use of the ToC templates. I assume you have custom templates?

Start screen: You can start `wxchordpro` with a file name or a directory name and it will take you straight into the editor or builder. (This is what happens when you doubleclick on a ChordPro document.)

I share your opinions on File > New. Can we do better? Without disturbing the current docs too much?

Rabbit hole: You are very right! Only title and subtitle are by default visible in the output. Same for the 'New Song' dialog.

**Comment by sciurius** [comment]
Cover page: When you add `--title` (and optionally `--subtitle`) to the cli you get the cover page, provided there is a ToC (use `--toc` to force). Easier for experimenting.

**Comment by sciurius** [comment]
When you are in the soingbook builder, doubleclicking on one of the songs will open the editor with that song, and there is a nice back-button to go back to the builder.

**Comment by gwyndaf** [comment]
> I share your opinions on File > New. Can we do better?

I think maybe two things, which shouldn't(?) be a big upheaval:
* Add File > Start Screen as a separate submenu item, maybe at the top, with separator.
* Given new functionality, perhaps File > New should bring up the new 'New Song' dialog (form with Title, Subtitle, Artist etc.), as currently happens when 'New song' is selected from the Start Screen. I think that's pretty consistent with things like MuseScore, and it's a quick `<Esc>` press for anyone who just wants the editor screen.

Yes, as you picked up, I realised the 'New Song' dialog also presents Key as 'core' metadata, like the menu. It's a nice touch to allow additional user-selected (drop-down) metadata variables, but I think 'Key' should simply be in the drop-down list, not fixed.

So, we'd have three (or two?) drop-downs for extra, user-selected metadata. How many there are doesn't matter much - there might always be a point at which users have to add further metadata manually, and we lose ease of use if the 'New Song' dialog is too complex.

What should those three metadata drop-downs default to? I wonder if a useful guiding principle is "What get printed at the top of a score's title page", rather than on the score stave itself. So, maybe Composer, Lyricist (or Copyright), Year? For me, that's all stuff that's _about the song_ rather than _how the song goes_, so 'more meta' than things like key or tempo.

Also, given the user-friendly focus, the names listed in the drop-down don't have to be the metadata variable names, so I'd suggest perhaps 'Time signature' or 'Time sig.' - it just helps avoid possible confusion with Duration.

Very nice touch being able to edit songs from within the songbook builder :)

> Cover page: When you add `--title` (and optionally `--subtitle`) to the cli you get the cover page

Thanks, that (and `toccover.cho` template) is useful to know. I was thinking there wasn't a CLI equivalent to the 'simple cover' GUI feature. That should make it easier to diagnose my configs, which are almost certainly the cause!

**Comment by sciurius** [comment]
A thought about creating a config. As it is now, a complete configuration file is created and somewhere in the docs there's a line saying to remove what you don't need. That's not inituitive.

When a user wants to create a configuration file she wants to adjust something, maybe a couple of things, probably (later) more. So I wonder if it is better to create a configuration file that is completely commented out. The user can then selectively remove the comments for items that she wants to adjust.
````
// Configuration for ChordPro
// 
// "include" takes a list of either filenames or preset names.
// include : [ guitar ]

// General settings, often changed by configs and command line.
// settings {
  // Titles flush: default center.
//  titles : center

...etc...

// }
````

Technically better, but still prone to errors. E.g., to adjust `titles`, you must remove the `//` before the `settings` line, the `titles` line and the closing brace line. 

I wonder if it would improve by un-nesting the items and use a different comment symbol:
````
// Configuration for ChordPro
// 
// Remove the leading # to access a configuration item.

// "include" takes a list of either filenames or preset names.
# include : [ guitar ]

// General settings, often changed by configs and command line.

// Chords parsing strategy.
// Strict (only known chords) or relaxed (anything that looks sane)
# settings.strict : true

// Titles flush: default center.
# settings.titles : center

...etc...
````

Now it is obvious (is it?) that you only need to remove the `#` before the `settings.titles` to access it.

Thoughts?

**Comment by gwyndaf** [comment]
That sounds like a great idea, so we have different comment markers for _descriptive_ (`//`) and _deactivated_ (`#`) lines.

It also means new, non-dev users can recognise the familiar `#` symbol as inactive lines, as in songs.

The nesting and braces convention has high potential to confuse novice users and be error-prone, so I'm all in favour of losing those to simplify/clarify. 

I haven't yet fully thought through and decided how I'd introduce config editing. I'm thinking to suggest putting only _changed_ items in custom configs, so users wouldn't necessarily need to copy and modify defaults. Nonetheless, the clearer the default config is to read, the more that helps the learning process.

One practical thought, for future consideration: for the user who wants to tweak their layout, but has no command-line experience, how might she access the default config to examine and/or modify it? That is, most GUI-based users are probably wary of navigating into any folders outside of 'My Documents'. This probably needs more thought, but your proposed config format seems a good basis, however users get to it.

**Comment by sciurius** [comment]
Would this be an option?

![scrot20241211163703](https://github.com/user-attachments/assets/7152d1c2-4221-4821-bf8a-95c148305234)

**Comment by gwyndaf** [comment]
Personally, I don't think the extra text is necessary and could be counterproductive, because:

* It locks the interface to existing built-in layouts, but future layouts could conceivably include Artist, Composer, Year etc.
* It's potentially off-putting to new users having to read a chunk of explanatory text and visit the website. I'm guessing some users will already have been reading the guide, while others will simply jump in and explore until they hit an obstacle/question, at which point they should find an answer in documentation. It's probably not a good precedent having aspects of the interface point to documentation quite so explicitly. That's a function better served by 'tooltips' on hover, but it's maybe unwise having that approach to build guidance into the interface at this stage of GUI development.
* For experienced users, it's 'pointless ink' that will probably be ignored after first use, so doesn't really earn its place in the interface.
* Users might get confused by "reference and administration purposes", and the more paranoid user might even think we're harvesting data on their songs!

I think it might be enough simply to:
* Split the data area of the form into two sections, _e.g._ with a divider line.
* Have a "Required" heading in the upper section, above Title, which simply puts "New Song" (as currently) by default.
* Have an "Optional" heading in the lower section, above the remaining fields, with Subtitle being the first of those. I don't think we then need to explain why they're optional, which can be covered in the tutorial or cookbook (I think "How can I output my song's key/tempo/etc?" might be a common question, so a natural one to include.

There's sense in making Subtitle a fixed entry, because it's a long-standing part of the format, and is output in the standard layouts. For other built-in metadata, personally, I'd be inclined to make them all drop-downs but with sensible default choices on the buttons. I'm not wholly sure what the best choices would be, and could be different for someone writing up a cover version from someone writing their own songs.

**Comment by sciurius** [comment]
Or drop this dialog completely and just insert `{title: New Song}`?

After all, that's all what is required to get going.

**Comment by gwyndaf** [comment]
I agree, that's simplest (and fastest for experienced users).

The form approach could create unrealistic expectations, of an interactive/guided interface for song editing tasks, which moves us away from the fundamentally text-based format. So, maybe best to get straight down to text editing from the outset.

If we want to offer similar ease of use and functionality as the New Song form, maybe adapt the Insert menu (currently just 4 metadata items as at _175), so it has a Song Information >  (or Song Metadata > )  entry (first menu item?), with all built-in metadata items listed at the next level (otherwise the Insert menu could quickly become very long)?

Just testing Insert, I notice a few things (Linux, _175):

1. The metadata directive currently get inserted _at the cursor point_, which could be immediately after or even within an existing directive or song line, which would be invalid. Perhaps new directives should automatically be inserted on a new line (after the cursor position) if the current line isn't empty. Inserting on the line _after_ cursor position would mean if I Insert > Title then Insert > Key, they appear in that order.
2. Similar thing for inserting sections: happens at cursor position, rather than in new line/s.
3. Insert Subtitle doesn't seem to work, _i.e._ nothing is inserted.

**Comment by sciurius** [comment]
@Desbeers I've removed the 'New Song' panel and the menu entries for insert key and artist due to the ['rabbit hole'](https://www.chordpro.org/chordpro/faq-pagetitles/).

**Comment by sciurius** [comment]
@gwyndaf I've fixed the insert issues in _202.

**Comment by gwyndaf** [comment]
Great, that menu approach seems good, so Insert covers just the most essential directives for basic song layout.

I'm also pleased to say I can now build the GUI version on Linux, which I couldn't before (for Wx-related reasons I don't really understand).

Issues with inserting directives at cursor position seem much improved, but still occur occasionally (Linux, v. _202). Starting with an empty document, if I `Insert > Title` (cursor inside close brace), then `Insert > Chorus`, `{start_of_chorus}` isn't now inserted at the cursor position, but is inserted immediately after `{title: }`, _i.e._ on the same line.

It seems a bit intermittent, though. At other times, with cursor inside `}`, the next inserted directive _does_ start on the next line. Odd.

**Close Section** is an interesting addition and potentially a quick way to add `{end_of_...}` directives _but_ a couple of issues:

**1)** It doesn't check whether the current section has already been closed, and could create a duplicate `{end_of_...}` directive. For novice users, that's potentially more confusing than helpful - if they've inserted a section with the GUI menu, they'll already have the relevant `{end_of_}` directive, and the cursor positioned ready to populate the section. I'm guessing it currently looks backward for a `{start_of_}`, but not for any corresponding `{end_of_}`.

**2)** There's also a semantic risk that the menu arrangement/wording could imply it's a step in inserting a section, _i.e._ 1) insert a section, then 2) close the section. So it may be that presenting the function in this way is counter-productive for novices. Maybe clearer if it's a) separated from the other Insert section options with a divider line, and b) reworded as "Close current section". Or maybe remove it altogether (for now).

The function (in some form) seems more useful for 'hands-on' users who manually type `{start_of_}` directives, but I wonder if there's something incongruous about providing menu actions (hand-on-mouse) to complete typing actions (hand-on-keyboard)? 

Maybe the logical resolution is to add a keyboard shortcut (whether or not we retain a menu entry), so it's more ergonomic for hands-on-keyboard users? It's not really a standard GUI operation, so I don't know if there's any natural choice of shortcut.

**3)** It's also suffering the 'inserted at cursor position' issue, in its original form, _i.e. inserted exactly at cursor position, even it that's inside `}`.

**Comment by Desbeers** [comment]
@gwyndaf
Unlike you, with your fancy words, I *contributed* to the current documentation. You promised to ’take the user documentation role’ but so far I did it.

Lets try to say this polite and friendly and become fiends.

- You are a bit late with debugging the new GUI and that is another issue.
- A good GUI does not need documentation, the *config* does.
- Deliver what you promised. Write documentation.

Ok, shall we shake hands?

Because we both have the same goal.

Friends?

**Comment by gwyndaf** [comment]
@Desbeers Of course. I'm quite practical and assume that all collaborators are equally focused on the job at hand. I'm sure we all want to move ChordPro forward, effectively and efficiently (subject to our other commitments), so I don't see any argument.

To clarify my contribution: words (and communication and education) are what I do professionally. I can do fancy if necessary, but what's usually needed is words that are clear and relevant for the target audience. I don't code, but value ChordPro, so I offered to contribute what skills I have that could serve the project (along with some testing, feature ideas and insight into non-technical minds).

My offer at the start of this github discussion meant that if anyone else raised an interest in contributing to the guidance, we could work together in a planned and coordinated fashion. If two people do the same work without communicating that, it's a waste of someone's time. I don't recall you offering to write a user's guide, so I didn't know you were. I just knew you'd developed a new GUI for MacOS.

When I made the offer, that didn't imply "do it now" as an immediate start or early completion, which you seem to have assumed. My aim was to begin once another project of mine was out of the way, which has happened more recently. Until Johan's mention this week of a Christmas release, I wasn't aware of any project timescale, so was basically working on my own timescale.

So, documentation has been progressing gradually as a background (thinking and outlining) job, to shape the novice user's journey. You haven't seen that yet, but I've said I'll do it, and I'm doing it. It just isn't in the form of sensible words and pictures yet.

_Are you saying you want to take on the responsibility for user documentation?_

I'm perfectly happy if that's what you want to do, and will write off the work I've already done. I'm don't like a competitive free-for-all, which never seems like a good use of time, and my time is limited, so you're welcome to take on the documentation responsibility if you want.

I _was_ aiming to devote some concentrated time to documentation next week, so please let me know by Sunday whether or not you want to take on that documentation responsibility instead of me?

**Comment by sciurius** [comment]
Nick, watch your language. Everyone is doing his best in his own way.

As far as I'm concerned there are multiple projects on the way. With help of Nick and his solid macOS user background I've made a complete new GUI. We would like to release this for Christmas.
For the release, the 'getting started' documentation must be made up-to-date. Nick is working on that.
He is also working on a new landing page that is easier for us to control and maintain.

Alyn, as I understood, is focussing on a new and revised documentation set, including with tutorials and cookbooks. That's a longer project and is not required for the Christmas release.

I have asked Alyn and Herbert to take a look at the new GUI, to straighten out user issues before we (beta) release it to the forum. Their feedback is very much appreciated and has already led to some crucial (fortunately minor) changes.

**Comment by Desbeers** [comment]
@gwyndaf
Oh no, my apologies. English is not my native language. I’m Dutch and we seems to talk rude while for us, it normal. I’m a seaman, we talk even *ruder*.

I hope you will accept my apologies. It wast mend to be constructive .

Leaning every day. Friends?

**Comment by AnniePerson** [comment]
With 6.60.194 the file from my previous post is rendered correctly.
I have to bring up again a feature request that I posted long ago: as soon as standard notation like ABC is involved it is essential that we can choose between transpose up and transpose down.
![AfterYouveGone_6 60 194](https://github.com/user-attachments/assets/8929e87a-ef55-45c4-b426-bb2e63a8d2c7)
![AfterYouveGone_6 60 194_Bb-G](https://github.com/user-attachments/assets/3b511c66-b3b0-42ef-89ec-d98d1bd0a459)

**Comment by AnniePerson** [comment]
btw. I think test results and feature requests should better be in a different thread. Maybe my fault.

**Comment by gwyndaf** [comment]
@Desbeers Really no problem, Nick. And I admire that the Dutch grasp of English is way better than my grasp of Dutch (although I'm actually Welsh)! :)

For practical purposes, if you've already documented the new GUI, perhaps we use that to accompany the Christmas release?

@sciurius That then frees up my time in the next week or two to test/review GUI iterations, especially now I can build the Wx version myself (for Linux at least: building for Windows is beyond me).

Revised user guidance will continue to develop, but I won't aim for Christmas, even with a rough GUI guide.

**Comment by gwyndaf** [comment]
@AnniePerson In fairness, both documentation and GUI should enhance the user experience, so it's a blurry boundary!

I'll start a fresh thread once I've got rationale and practical development of documentation into a form that's ready to share for review and discussion. So far, I've been gathering scraps of thoughts, ideas and structure, which need some dedicated time to shape, hopefully over the Christmas break.

**Comment by Desbeers** [comment]
@gwyndaf
Thanks, makes me feel better :-)

The new GUI documentation is already on https://www.chordpro.org/beta/chordpro-getting-started-gui/

I’m not good at documenting an application. I just updated the current docs to be *correct* and visual pleasant. I’m more a (GUI) programmer than a writer.

I’ll leave the next step of documentation up to you.

**Comment by gwyndaf** [comment]
Great. On the 'directive on same line' issue, I've not yet been able to break that with _204.

'Close current section' menu presentation seems clearer. It's still giving me duplicated `{end_of_}` directives, inserted at cursor position (not new line). As you say, such duplication is clearly visible so not a huge issue, but I know some people don't have a feedback process, and just 'execute the steps' without monitoring the results. Hopefully the overlap with musicians is small!

I notice `File > New` functions differently according to whether it's run from the Start Screen or Editor. In Start Screen it opens a new document with boilerplate `{title}`. In Editor, it shows the Start Screen, so the same menu operation has different effect depending on context. 

Perhaps that's confusing? A couple of possible alternatives:

1. When invoked from Editor, have the menu entry titled `File > Start Screen` instead of `File > New`
2. Have `File > New` open Editor with boilerplate new song in all cases, but add a `File > Start Screen` menu item to invoke the Start Screen` (which will do nothing / be grey if already at Start Screen).

A thought on the 'big buttons' in Editor and Songbook Builder: are these clearer in the order `Preview`, `Settings`, `Messages`? `Preview` is what most users will want most of the time, so should maybe come first (assuming left-to-right reading). `Settings` maybe used less frequently, and `Messages` only if problems arise (or to 'see the workings'). But, this is an ease-of-use observation more than functionality, so maybe not a high priority.

**Comment by sciurius** [comment]
Proposed landing page: https://chordpro.org/landing.html .

**Comment by sciurius** [comment]
> A thought on the 'big buttons' in Editor and Songbook Builder: are these clearer in the order Preview, Settings, Messages? Preview is what most users will want most of the time, so should maybe come first (assuming left-to-right reading). Settings maybe used less frequently, and Messages only if problems arise (or to 'see the workings').
@Desbeers ?

**Comment by Desbeers** [comment]
In case of **ChordPro**, *Settings* are important for the *Preview*. The *Preview* gives *Messages*. So the order is natural to me.

I’m too much involved with the GUI prgogramming to have a *user* opinion.

**Comment by sciurius** [comment]
> Have File > New open Editor with boilerplate new song in all cases, ...

Please try _205.

**Comment by gwyndaf** [comment]
_205 works great with the `File` menu, and I think will be clearer for users.

**Comment by Desbeers** [comment]
@sciurius 

You are on the roll for documentation!

Details like https://github.com/ChordPro/chordpro/commit/9d9e7c3342c3fb0d098516de273d4f5959617b06 I appreciate very much.

I feel a bit guilty in a certain way. For no reason.

I did my best, a *good* GUI does not need documentation but I *really* like your documentation.

Why so much effort? Do you consider this a *showstopper*? 

I don’t.

But your are a better *‘application’* documenter than me for sure!

**Comment by sciurius** [comment]
> > Cover page: When you add `--title` (and optionally `--subtitle`) to the
> > cli you get the cover page
> 
> Thanks, that (and `toccover.cho` template) is useful to know. I was
> thinking there wasn't a CLI equivalent to the 'simple cover' GUI feature.
> That should make it easier to diagnose my configs, which are almost
> certainly the cause!

Please try dev _218. I've decoupled the cover page from the ToC so it
should produce consistent results. The template for the cover is looked up
in templates/cover.cho.

**Comment by gwyndaf** [comment]
Great, I can now get _218 producing a standard cover page, even with my custom configs, using both CLI and GUI (Linux).

_But_, to get it working, I had to fully remove the `contents` section from my own config. Even setting `omit : true` for all entries still prevents cover page generation. Does that suggest it's not fully decoupled from ToC processing?

Works fine when using default `-X` config, which suggests my config is confusing it, but I'm not sure how. I'll try to dig deeper, to see if I can narrow it down, but this is the offending config section in case you see something wrong.
```
    "contents" : [
		{ "fields"   : [ "songindex" ],
      "label"    : "Table of Contents (supplied order)",
      "line"     : "%{title} %{key_actual|(%{})|%{key|(%{})}} ",
      "pageno"   : "%{page}",
      "fold"     : false,
      "omit"     : true,
    },
    { "fields"   : [ "sorttitle" ],
      "label"    : "Contents by Title",
      "line"     : "%{title}<small>%{year|  (%{})}</small>%{artist|  —  %{}}<small>%{key_actual|  (%{})|%{key|  (%{})}}</small>",
      "pageno"   : "%{page}",
      "fold"     : false,
      "omit"     : false,
    },
    { "fields"   : [ "artist", "sorttitle" ],
      "label"    : "Contents by Artist",
      "line"     : "%{artist|%{}  —  }%{title} %{key_actual|(%{})|%{key|(%{})}}",
      "pageno"   : "%{page}",
      "fold"     : false,
      "omit"     : true,
    },
    { "fields"   : [ "key", "sorttitle" ],
      "label"    : "Contents by Key (roughly)",
      "line"     : "%{key_actual|%{} (orig. %{key})  –  |%{key}  –  }%{title} ",
      "pageno"   : "%{page}",
      "fold"     : false,
      "omit"     : true,
    },
    { "fields"   : [ "year", "sorttitle" ],
      "label"    : "Contents by Year",
      "line"     : "%{year|%{}  –  }%{title} ",
      "pageno"   : "%{page}",
      "fold"     : false,
      "omit"     : true,
    },
],
```

**Comment by gwyndaf** [comment]
After a bit more testing, it looks like the `[ "songindex" ]` ToC definition is causing trouble, even though omitted:
```
    { "fields"   : [ "songindex" ],
      "label"    : "Table of Contents (supplied order)",
      "line"     : "%{title} %{key_actual|(%{})|%{key|(%{})}} ",
      "pageno"   : "%{page}",
      "fold"     : false,
      "omit"     : true,
    },
```
If I delete that from my test config, cover gets generated ok. I've no idea why that is, but maybe it offers a clue?

**Comment by sciurius** [comment]
Almost... There was an entanglement where the cover template was not picked up when one of the toc items had omit set to true. This should be fixed in _219.

**Comment by gwyndaf** [comment]
Fantastic. _219 seems to work nicely.

Tested on both CLI and GUI (Linux).

**Comment by sciurius** [comment]
Great! I'll prepare for release.

**Comment by sciurius** [comment]
As of 6.060_230, creating a new config file will create a _template_ config, i.e. one with all settings commented out.
````
// Starter Configuration for ChordPro.
// 
// This configuration contains the majority of settings that may
// be of interest for beginning users.

// ALL CONFIGURATION SETTINGS ARE COMMENTED OUT WITH A LEADING "# ".
// IF YOU WANT TO CHANGE A CONFIGURATION SETTING, REMOVE THE "# " FIRST.

// Chords parsing strategy.
// Strict (only known chords) or relaxed (anything that looks sane)
# settings.strict : true

// Titles flush: default center.
# settings.titles : center
...
````

For command line use, there is a new option `--print-template-config`.

I'd say this is the preferred way to get going with custom configuration.

**Comment by gwyndaf** [comment]
I've started a new thread,  #490, to allow more focused discussion/feedback on the overall approach for proposed user documentation.

A separate thread for the main Tutorial outline will follow in the coming days.

**Comment by sciurius** [comment]
Excellent!

**Comment by gwyndaf** [comment]
Thinking for Tutorial now shared at #500 and open for feedback, before I proceed to drafting an outline of the linear structure.

**Comment by gwyndaf** [comment]
Topic outline for the Tutorial now [here](https://github.com/ChordPro/chordpro/issues/500#issuecomment-2569285249), and open for feedback, especially if you've got experience of trying to explain ChordPro concepts and usage to novice users.
